Implement returning payment hash (#1159)

Patti88 opened the pull request.

JssDWt commented on libs/sdk-bindings/src/breez_sdk.udl:
```diff
@@ -1,12 +1,12 @@
-     u64? htlc_minimum_msat;
-     u64? htlc_maximum_msat;
- };
+    string src_node_id;
+    string short_channel_id;
+    u32 fees_base_msat;
+    u32 fees_proportional_millionths;
+    u64 cltv_expiry_delta;
+    u64? htlc_minimum_msat;
+    u64? htlc_maximum_msat;
+};
```
This whitespace change is not needed

JssDWt commented on libs/sdk-bindings/src/breez_sdk.udl:
```diff
@@ -626,8 +625,9 @@ dictionary BitcoinAddressData {
 dictionary LnUrlPaySuccessData {
-    SuccessActionProcessed? success_action;
     Payment payment;
+    Optional<SuccessActionProcessed> success_action;
```
The success_action field doesn't need a change here? Does the order matter?
At least it should be `SuccessActionProcessed? success_action;`

JssDWt commented on libs/sdk-bindings/src/breez_sdk.udl:
```diff
@@ -918,7 +918,7 @@ interface BlockingBreezServices {
    [Throws=SdkError]
    SwapInfo? in_progress_swap();
-  
+
```
Whitespace change not needed

JssDWt commented on libs/sdk-bindings/src/breez_sdk.udl:
```diff
@@ -974,7 +974,7 @@ interface BlockingBreezServices {
 };
 
-namespace breez_sdk {  
+namespace breez_sdk {
```
Whitespace change not needed

JssDWt commented on libs/sdk-core/src/lnurl/pay.rs:
```diff
@@ -487,10 +488,11 @@ pub(crate) mod tests {
-            },
+            LnUrlPayResult::EndpointSuccess { data } => {
+                let expected_hash = inv.payment_hash().to_hex();
+                assert_eq!(data.payment_hash, expected_hash);
```
👍

JssDWt commented on libs/sdk-core/src/lnurl/pay.rs:
```diff
@@ -26,6 +26,7 @@ pub enum LnUrlPayResult {
 pub struct LnUrlPaySuccessData {
     pub payment: Payment,
     pub success_action: Option<SuccessActionProcessed>,
+    pub payment_hash: String, // Add payment_hash field
```
```suggestion
    pub payment_hash: String,
```

JssDWt commented on libs/sdk-core/src/models.rs:
```diff
@@ -659,6 +659,7 @@ pub enum PaymentStatus {
 #[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
 pub struct Payment {
     pub id: String,
+    pub payment_hash: String, // Explicitly adding the payment_hash field
```
```suggestion
    pub payment_hash: String,
```

Patti88 replied:
Thanks for your feedback!  I see your point about the payment_hash, but it's actually supposed to be there.  It's how we keep track of each payment.

JssDWt reviewed (changes requested):
Thanks for the PR!

There's quite a few changes unrelated to the PR. Some I've listed in comments. The following files shouldn't need any changes, can you remove those changes please?

Please remove the changes to
- .gitignore
- libs/Cargo.lock
- libs/sdk-bindings/bindings-android/buildForJitpack.sh
- libs/sdk-bindings/bindings-android/gradle/wrapper/gradle-wrapper.jar
- libs/sdk-bindings/bindings-android/gradlew
- libs/sdk-bindings/bindings-android/gradlew.bat
- libs/sdk-bindings/bindings-csharp/src/.gitignore
- libs/sdk-bindings/bindings-csharp/src/Breez.Sdk.csproj
- libs/sdk-bindings/bindings-csharp/test/.gitignore
- libs/sdk-bindings/bindings-csharp/test/Program.cs
- libs/sdk-bindings/bindings-csharp/test/sdk-cs-demo.csproj
- libs/sdk-bindings/bindings-kotlin-multiplatform/gradle/wrapper/gradle-wrapper.jar
- libs/sdk-bindings/bindings-kotlin-multiplatform/gradlew
- libs/sdk-bindings/bindings-kotlin-multiplatform/gradlew.bat
- libs/sdk-bindings/tests/bindings/csharp/.gitignore
- libs/sdk-bindings/tests/bindings/csharp/Program.cs
- libs/sdk-bindings/tests/bindings/csharp/sdk-cs-demo.csproj
- libs/sdk-common/Cargo.toml
- libs/sdk-common/build.rs
- libs/sdk-common/src/breez_server.rs
- libs/sdk-common/src/buy/moonpay.rs
- libs/sdk-common/src/fiat.rs
- libs/sdk-common/src/grpc/breez.rs
- libs/sdk-common/src/grpc/mod.rs
- libs/sdk-common/src/grpc/proto/breez.proto
- libs/sdk-common/src/model.rs
- libs/sdk-core/src/breez_services.rs
- libs/sdk-core/src/persist/send_pays.rs
- libs/sdk-core/src/swap_out/reverseswap.rs
- libs/sdk-react-native/android/gradle/wrapper/gradle-wrapper.jar
- libs/sdk-react-native/android/gradlew
- libs/sdk-react-native/android/gradlew.bat
- libs/sdk-react-native/example/android/gradle/wrapper/gradle-wrapper.jar
- libs/sdk-react-native/example/android/gradlew
- libs/sdk-react-native/example/android/gradlew.bat

Patti88 commented:
Hi @JssDWt, thank you for the review,  I have fixed the success_action type, removed the extra payment_hash field in lnurl/pay.rs, and confirmed that the payment_hash in models.rs is correct. I've also reverted the changes to the unrelated files you pointed out.

JssDWt commented:
@Patti88 If you look at the 'files changed' tab, there's still 38 changed files. Can you bring that down to only the relevant files? When that's done please squash the commits in a single commit. You could use Github Desktop for example, it has a nice diff viewer and allows you to squash commits by selecting them and right-clicking.

roeierez commented:
The diff is still very large and not relevant.
Also @Patti88 I see your comment here: https://github.com/breez/breez-sdk-greenlight/issues/1159#issuecomment-2653170824
The comment was about reverse swap and this PR is about lnurl. Was that the intention?
Asking also because the Payment object already contains the payment hash (inside PaymentDetails when relevant) so I am not sure there is a need to do something for lightning/lnurl payments.

JssDWt commented:
I see the diff has been much improved. Now I can see the trees in the forest. Let me review the code.

You're creating a fix for [this issue](https://github.com/breez/breez-sdk-greenlight/issues/1159).
The goal in that issue is to return the payment hash from the `claim_reverse_swap` function.

That function is [here](https://github.com/breez/breez-sdk-greenlight/blob/bb68587b221cd1c4a0471b7bc98b3cd0082144f7/libs/sdk-bindings/src/breez_sdk.udl#L953) in the udl file. It should return a response object, rather than `void`. 

The corresponding function implementation is [here](https://github.com/breez/breez-sdk-greenlight/blob/bb68587b221cd1c4a0471b7bc98b3cd0082144f7/libs/sdk-core/src/breez_services.rs#L893-L898). That should return that new object. You can retrieve the payment hash by hashing the [preimage](https://github.com/breez/breez-sdk-greenlight/blob/bb68587b221cd1c4a0471b7bc98b3cd0082144f7/libs/sdk-core/src/models.rs#L146), or decoding the [invoice](https://github.com/breez/breez-sdk-greenlight/blob/bb68587b221cd1c4a0471b7bc98b3cd0082144f7/libs/sdk-core/src/models.rs#L154). Decoding the invoice is probably the preferred way.

Currently this PR changes the response of the `lnurl_pay` function, but the payment hash is already included in the returned `payment` there.

I would suggest to start fresh with this PR, implementing the required changes for the mentioned issue.

JssDWt commented:
Closing this for inactivity. Feel free to reopen.
